kpi-tracker/src/main.rs

```diff
@@ -79,117 +84,396 @@ struct BlockDetails {
 struct AccountDetails {
     /// Whether an account was created as an initial account or not.
     is_initial: bool,
-    /// Foreign key to the block in which the account was created.
-    block_hash: BlockHash,
 }
 
 /// Holds selected attributes of an account transaction.
 #[derive(Debug)]
 struct TransactionDetails {
     /// The transaction type of the account transaction. Can be none if
     /// transaction was rejected due to serialization failure.
-    transaction_type: Option<TransactionType>,
-    /// Foreign key to the block in which the transaction was finalized.
-    block_hash:       BlockHash,
+    transaction_type:   Option<TransactionType>,
     /// The cost of the transaction.
-    cost:             Amount,
+    cost:               Amount,
     /// Whether the transaction failed or not.
-    is_success:       bool,
-}
-
-/// Holds selected attributes of a contract module deployed on chain.
-#[derive(Debug)]
-struct ContractModuleDetails {
-    /// Foreign key to the block in which the module was deployed.
-    block_hash: BlockHash,
+    is_success:         bool,
+    /// Accounts affected by the transactions.
+    affected_accounts:  Vec<CanonicalAccountAddress>,
+    /// Contracts affected by the transactions.
+    affected_contracts: Vec<ContractAddress>,
 }
 
 /// Holds selected attributes of a contract instance created on chain.
 #[derive(Debug)]
 struct ContractInstanceDetails {
     /// Foreign key to the module used to instantiate the contract
     module_ref: ModuleRef,
-    /// Foreign key to the block in which the contract was instantiated.
-    block_hash: BlockHash,
 }
 
-/// Represents a relation between an account and a transaction
-#[derive(Debug, Hash, PartialEq, PartialOrd, Ord, Eq)]
-struct TransactionAccountRelation {
-    account:     CanonicalAccountAddress,
-    transaction: TransactionHash,
+/// Map of accounts indexed by canonical address
+type Accounts = HashMap<CanonicalAccountAddress, AccountDetails>;
+/// Map of transactions indexed by transaction hash
+type AccountTransactions = HashMap<TransactionHash, TransactionDetails>;
+/// Set of contract modules references
+type ContractModules = HashSet<ModuleRef>;
+/// Map of contract instances indexed by contract address
+type ContractInstances = HashMap<ContractAddress, ContractInstanceDetails>;
```

Comment:
I strongly recommend using BTreeMap and BTreeSet. HashMaps have unspecified iteration order, and while that might not matter in most cases, it does mean we have potential nondeterminism in the way the data is stored, and that will possibly lead to some nondeterminism in queries as well.

Reply:
Makes sense, will update!